### ISSUE TITLE
Easier cloning / building

### ISSUE THREAD

**Issue by Quincunx271** [CSUS-LLVM/OptSched#48]
Currently, [BUILD.md](https://github.com/CSUS-LLVM/OptSched/blob/master/BUILD.md) and [documentation/flang/README.md](https://github.com/CSUS-LLVM/OptSched/blob/master/documentation/flang/README.md) list a large number of steps which require manual interaction interspersed with waiting for compilation to finish. Ideally, we could just do a single download step and a single build step (possibly with a single prior configure step).

I made a master repository which allows for this via git submodules: https://github.com/Quincunx271/OptSchedProject . A recursive git clone grabs all the source code needed plus OptSched itself, and there's a super-build CMakeLists.txt to kick off the build for everything. You can also clone only LLVM + OptSched if flang is not desired.

The idea is to have CSUS-LLVM/OptSchedProject and associated git repositories to make setting up OptSched easier. The associated repositories are the patched versions of llvm-project and flang-llvm.

I tried testing it on grace2, but the cloning of llvm-project appears to freeze grace2 and prevent new ssh sessions.

**TL;DR**: The future of setting up and building OptSched:

```
$ git clone --recursive https://github.com/CSUS-LLVM/OptSched-SuperProject.git
$ cd OptSched-SuperProject && mkdir build && cd build
$ cmake <args> ..
$ ninja
```

-----

Naming is hard. Any suggestions on what to name this? "OptSchedProject" is not that good of a name.

**Comment by vangthao95** [comment]
How about a configure script which will download the required packages/repos and run the commands? This way we don't need two repo and we can simply add options if we want to target x86 or the amdgpu. For x86 you will need flang whereas on the amdgpu you will need ROCm.

**Comment by Quincunx271** [comment]
I thought about that, but I'm reluctant. A script to download the required packages and the like is hard to get working right. In my experience, every such script I've written is great if it works, but there's often something that stops working for some reason or another, leaving things in some partially-formed state.

The advantage of using git is that it's git; it's solved a lot of problems for you. It's a lot easier to keep a git repo cloneable than it is to write a script to handle downloading and configuring for you.

I recognize that this does not handle the amdgpu and ROCm because I don't know how to set this up when using the `repo` tool.

**Comment by Quincunx271** [comment]
Upon further thought: the main issue I was concerned with is robustness. I think there's a way to script it and remain robust: using CMake. This would be minor changes from what I already have, but would not require separate repositories.

The one thing I'm unsure of is how the script would handle placing OptSched inside of LLVM when the script doesn't get to clone OptSched because it's part of OptSched. I think CMake does not handle symlinks well, so it's a bit awkward.

**Comment by vangthao95** [comment]
Actually, now that I think about it, OptSched is meant to be an open-source plugin for LLVM. Others may not be interested in flang or ROCm at all so we should still keep the steps in BUILD.md. We can probably remove the steps for MacOS since we have never done tests on MacOS and cannot guarantee everything works on it. The script or master repository should be optional and will most likely only be used internally by the team to set up our benchmarking environment. If a script is too hard to create then I wouldn't mind a private internal master repository with possibly different branches for x86 and the amdgpu target

@kerbowa any opinions on this?

**Comment by Quincunx271** [comment]
I've got a script mostly set up. It's a subdir of this repo (OptSched): cmake/superbuild/CMakeLists.txt. Building with CMake downloads all the dependencies including llvm and sets everything up, in theory at least (still waiting for it to compile)

**Comment by Quincunx271** [comment]
I have a branch which supports this script-model superbuild: https://github.com/Quincunx271/OptSched/tree/superbuild

It would be possible to create another script to support the amdgpu build.

**Comment by vangthao95** [comment]
If you can confirm that it works then I think it would be a great addition to have.

**Comment by kerbowa** [comment]
I think this is a needed and we should add it to the main CSUS-LLVM project.

A good alternative to git submodules would be using [repo](https://gerrit.googlesource.com/git-repo/), like we use for ROCm. We could have manifests for the different configurations we want to support. This would avoid the need to fork the repos we don't modify.

My suggestion:
CSUS-LLVM/llvm-project -- since we modify LLVM in some builds we can have those changes in this fork
CSUS-LLVM/OptSched -- here we can put the manifests for the different build types as well as build scripts

git submodules may work just as well, but I'm not familiar enough with them to say for sure.

Here is an example repo manifest for the ROCm opencl driver.
https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime/blob/roc-2.3.x/opencl.xml

**Comment by Quincunx271** [comment]
@vangthao95 I can confirm that the super-build CMakeLists.txt worked on my machine. I would still need to write documentation and think about polishing the build a bit.

@kerbowa Do you think it's better to have the repo / git submodule approach or to have a super-build script? Either approach works (I definitely have implementations of both).

Benefits I see of using the git submodules approach:

 - We can manage our changes to LLVM / flang-llvm in separate git repositories rather than patch files
 - Clearly separated "get source" and "build" steps. The CMake super-build script intermixes the two
 - CMake 3.4 can be supported (the super-build script requires 3.7+).

Benefits I see of using the super-build script approach:

 - You clone OptSched directly rather than a super-repo.
 - Only one repo is necessary rather than three (OptSched, llvm-project, flang-llvm)

**Comment by kerbowa** [comment]
I like the submodule/repo approach it seems more standard. Doing version control with cmake seems unusual to me. I think either approach will work fine though.

LLVM uses a monorepo. ROCm uses repo. We used to use a bash script for this: https://github.com/CSUS-LLVM/dragonegg-OptSched/blob/master/CompileLLVMandDRAGONEGG.sh

We need updated cmake scripts either way even if we don't clone the dependencies that way.

**Comment by Quincunx271** [comment]
I also tend to like the submodule approach. I don't know repo, though, so it will be longer before I can get an amdgpu setup, whereas the CMake approach would be straightforward despite using repo.

My biggest concern is that this git submodules method seems to break Grace2; it's been offline since I attempted to clone the master Super-Build repo. It would probably be able to handle the CMake script, though.

**Comment by kerbowa** [comment]
@Quincunx271 Do you know with submodules if you can have different repos per-branch. I.e. only clone ROCm repos when you are on the AMDGPU branch or similar.

**Comment by Quincunx271** [comment]
I'm 99% sure this is possible. I believe submodules work solely through the .gitmodules file, which, like any other file, can differ between branches.

**Comment by kerbowa** [comment]
That sound good to me then we can skip repo.
So we should fork LLVM only right, or is there some other non-optsched project we modify?

**Comment by Quincunx271** [comment]
We also modify flang-llvm.

**Comment by Quincunx271** [comment]
I've learned that git submodules requires that the super-project is updated every time OptSched is updated; otherwise, you check out an old commit even when doing a fresh git clone. You have to call a command to update the submodules in the super-project every time, unless there's something out there which I'm not aware of.

This makes me reluctant to say we should go for the submodules approach. I think it may be fine to have submodules for llvm or things which we rarely change, but the fact that it tracks commit SHAs and not arbitrary branches/tags makes me reluctant to go this way.

https://softwareengineering.stackexchange.com/a/342169

**Comment by kerbowa** [comment]
Aren't these the reasons that people prefer repo? It does work with tags or just whatever commit hash is in our manifest.

**Comment by Quincunx271** [comment]
It looks like it. I have no familiarity with repo, though.

**Comment by kerbowa** [comment]
When building targeting flang do you use the same flang-llvm to build clang and the c/c++ support as well, or is it a separate copy of LLVM.

**Comment by Quincunx271** [comment]
I've set it up as a separate copy. Could it be reused? I saw the URL was
different and assumed there were slight differences

**Comment by kerbowa** [comment]
There are differences. I'm asking what clang+optsched+llvm you use when you are also building flag.

Presumably you could build clang+optsched with the version of LLVM that you also use to build flang. Is this whats currently done or what.

**Comment by Quincunx271** [comment]
No, currently it's two entirely separate LLVM builds. One is the LLVM which
wraps OptSched, and the other is flang-llvm. I don't know enough about the
differences between the two to know if this would work, but it would be
easy enough to set up.

**Comment by kerbowa** [comment]
Do you build two separate OptSched libraries as well?

**Comment by Quincunx271** [comment]
No. Perhaps I misunderstood the documentation on building flang; I didn't
see anything saying to do that.

**Comment by kerbowa** [comment]
What I'm getting at is the differences between LLVM versions cannot be that great if the same build of OptSched is compatible with both versions.

Another TODO along these lines is having OptSched built as a static part of LLVM and not as a plugin. An integrated build should be an option for complex stacks where specifying plugin options isn't easy. For example I had to do this when using our scheduler with TensorFlow. I was thinking about how the build system will effect this. I couldn't easily find a way to have cmake give this option and was forced to embed the OptSched code within LLVM and do some tricks to get it to link properly. 

To move forward with this issue we need to agree on a name right? I think the org CSUS_LLVM should be renamed to OptSched. Next we need a name for the superproject. Maybe just steal from LLVM again and go with OptSched-project?

**Comment by kerbowa** [comment]
Oh right that's exactly what you already named it. That sounds fine to me.

**Comment by Quincunx271** [comment]
I was thinking of renaming it OptSched-SuperProject to try to more clearly
distinguish it from OptSched

**Comment by Quincunx271** [comment]
Due to my lack of experience with repo, if I was to proceed with this, I would proceed with the CMake super-build script rather than the submodules approach. I don't think it would be efficient use of my time to learn repo well enough to set this up. Using ExternalProject in CMake for this sort of thing is common enough, and it removes the need for the user to install yet another tool (repo).

**Comment by kerbowa** [comment]
Okay, so will you make changes in the main org CSUS_LLVM?

**Comment by Quincunx271** [comment]
Yes, I'll do that

**Comment by Quincunx271** [comment]
I noticed that the flang build instructions don't seem to mention any OptSched specific things other than the patches. Should OptSched by inside the flang-llvm's projects/ subdirectory as well?

**Comment by Quincunx271** [comment]
I'm planning on having us fork llvm-project and flang-llvm with the patches applied. Can I get permissions for https://github.com/CSUS-LLVM/llvm-project/ ?

**Comment by Quincunx271** [comment]
And can we also get a fork of https://github.com/flang-compiler/llvm/ ?

**Comment by vangthao95** [comment]
OptSched does not need to be inside flang-llvm if you built already built OptSched with the regular llvm.

**Comment by kerbowa** [comment]
> OptSched does not need to be inside flang-llvm if you built already built OptSched with the regular llvm.

Then presumably the inverse is true as well and you only really need to build LLVM once when using 6.0 clang/flang.
